frame/assets/src/functions.rs

```diff
@@ -333,7 +348,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		ensure!(account.balance.is_zero() || allow_burn, Error::<T, I>::WouldBurn);
 		ensure!(!account.is_frozen, Error::<T, I>::Frozen);
 
-		T::Currency::unreserve(&who, deposit);
+		let mut refund_to: T::AccountId = who.clone();
+		if let Some(depositor) = account.depositor {
```

Comment:
should be depositor.take() here, otherwise if it wont be removed below, the account still have a depositor with no deposit.
also in the else branch below we need to update the `Account`, now it should not really enter this branch, but if `dead_account` logic changes, this will be wrong, and will open a way for endless refunds.